jupyter-extension/jupyterlab_pachyderm/pfs_manager.py

```diff
@@ -214,16 +219,325 @@ def get(self, path, content=True, type=None, format=None) -> dict:
+        #  - if the dir is a parent of a file or directory that's mounted (i.e. if the path exists)
+        fileinfo_path = Path(self._FILEINFO_DIR, path)
+
+        if fileinfo_path.exists():
+            if fileinfo_path.is_file():
+                with open(fileinfo_path, "r") as f:
+                    fileinfo = pfs.FileInfo().from_json(value=f.read())
+                return fileinfo.file_type == pfs.FileType.DIR
+            return True
+
+        for dir in fileinfo_path.parents:
+            if str(dir) in self._dirs:
+                file = self._get_file_from_path(path=path)
+                try:
+                    fileinfo = self._client.pfs.inspect_file(file=file)
+                except grpc.RpcError as err:
+                    if err.code() == grpc.StatusCode.NOT_FOUND:
+                        return False
+                    raise err
+                return fileinfo.file_type == pfs.FileType.DIR
+
+        return False
+
+    def exists(self, path):
+        fileinfo_path = Path(self._FILEINFO_DIR, path)
+
+        if fileinfo_path.exists():
+            return True
+
+        for dir in fileinfo_path.parents:
+            if str(dir) in self._dirs:
+                file = self._get_file_from_path(path=path)
+                try:
+                    fileinfo = self._client.pfs.inspect_file(file=file)
+                except grpc.RpcError as err:
+                    if err.code() == grpc.StatusCode.NOT_FOUND:
+                        return False
+                    raise err
+                return True
+
+        return False
+
+    # translate a Path in local storage to a path in the contents FS
+    def _translate_path(self, path: Path) -> str:
+        if not path or len(Path(path).parts) <= 2:
+            return "/"
+        return str(Path(*Path(path).parts[2:]))
 
+    def _get_model_from_fileinfo(
+        self, fileinfo: pfs.FileInfo, path: Path, content: bool, type: str, format: str
+    ) -> dict:
+        model = _create_base_model(path=path, fileinfo=fileinfo, type=type)
         if content:
             if model["type"] == "file":
-                self._get_file_model(model=model, path=path, file=file, format=format)
+                _get_file_model(
+                    client=self._client,
+                    model=model,
+                    path=path,
+                    file=fileinfo.file,
+                    format=format,
+                )
             else:
-                self._get_dir_model(model=model, path=path, file=file)
+                _get_dir_model(
+                    client=self._client, model=model, path=path, file=fileinfo.file
+                )
+        return model
 
+    def _get_parent_dir_model(self, local_path: Path, content: bool) -> dict:
+        model = {}
+        model["name"] = local_path.name
+        model["path"] = self._translate_path(path=local_path)
+        model["created"] = self._mount_time
+        model["last_modified"] = self._mount_time
```

Comment:
Should created/last_modified time be mount time or pachyderm commit time?

Reply:
i have it implemented this way because of the way we mount datums. since we only get back a list of datuminfo each containing a list of files, if a directory only "exists" in the current state of the contents manager because one of its children is mounted, we have no data available for any commit info on that directory. it could potentially be an option to traverse down to where the children of the directory are being mounted to find out those commit times instead, but that would complicate the implementation and doesn't actually seem all that useful to me, given we aren't mounting the entire directory and may be missing commit data from other files contained therein.